src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -21628,6 +21628,146 @@ struct ConcatenateBroadcastInDim
+      }
+    }
+
+    // Check if rhs base is rotate of lhs base
+    if (dim == -1) {
+      if (auto rotateOp =
+              rhsInfo.base.template getDefiningOp<enzymexla::RotateOp>()) {
+        if (rotateOp.getOperand() == lhsInfo.base) {
+          dim = rotateOp.getDimension();
+          amount = rotateOp.getAmount();
+          operand = lhsInfo.base;
+          flippedOrdering = true;
+        }
+      }
+    }
+
+    if (dim == -1)
+      return failure();
+
+    auto wrapOp = rewriter.create<enzymexla::WrapOp>(op.getLoc(), operand,
+                                                     amount, 0, dim);
+
+    SmallVector<int64_t> windowDims(outRank, 1);
+    windowDims[dim] = 2;
+    SmallVector<int64_t> windowDilations(outRank, 1);
+    windowDilations[dim] = amount;
+
+    auto zeroType = RankedTensorType::get({}, outType.getElementType());
+    auto zero = rewriter.create<stablehlo::ConstantOp>(
+        op.getLoc(), zeroType, cast<ElementsAttr>(makeAttr(zeroType, 0)));
+
+    auto reduceWindowOp =
+        rewriter.replaceOpWithNewOp<stablehlo::ReduceWindowOp>(
+            op, TypeRange(outType), ValueRange(wrapOp), ValueRange(zero),
+            rewriter.getDenseI64ArrayAttr(windowDims), DenseI64ArrayAttr(),
+            DenseI64ArrayAttr(), rewriter.getDenseI64ArrayAttr(windowDilations),
+            DenseIntElementsAttr());
+
+    {
+      auto *block = rewriter.createBlock(&reduceWindowOp.getBody());
+      auto argType = RankedTensorType::get({}, outType.getElementType());
+      block->addArgument(argType, op.getLoc());
+      block->addArgument(argType, op.getLoc());
+      rewriter.setInsertionPointToStart(block);
+
+      Value lhsVal = block->getArgument(0);
+      Value rhsVal = block->getArgument(1);
+
+      if (lhsInfo.multiplier.has_value()) {
+        lhsVal = rewriter.create<stablehlo::MulOp>(
+            op.getLoc(), lhsVal,
+            rewriter.create<stablehlo::ConstantOp>(
+                op.getLoc(), lhsVal.getType(),
+                cast<ElementsAttr>(SplatElementsAttr::get(
+                    cast<RankedTensorType>(lhsVal.getType()),
+                    ArrayRef(lhsInfo.multiplier.value())))));
+      }
+
+      if (rhsInfo.multiplier.has_value()) {
+        rhsVal = rewriter.create<stablehlo::MulOp>(
+            op.getLoc(), rhsVal,
+            rewriter.create<stablehlo::ConstantOp>(
+                op.getLoc(), rhsVal.getType(),
+                cast<ElementsAttr>(SplatElementsAttr::get(
+                    cast<RankedTensorType>(rhsVal.getType()),
+                    ArrayRef(rhsInfo.multiplier.value())))));
+      }
+
+      if (flippedOrdering) {
+        lhsVal, rhsVal = rhsVal, lhsVal;
+      }
+
+      rewriter.create<stablehlo::ReturnOp>(
+          op.getLoc(),
+          ValueRange(rewriter.create<OpTy>(op.getLoc(), lhsVal, rhsVal)));
+    }
+
+    return success();
+  }
+
+private:
+  struct OperandInfo {
+    Value base;
+    std::optional<Attribute> multiplier;
```

Comment:
Why not just store std::optional<Value>, so you don't need to rematerialize the constant?

Reply:
sorry std::optional<Value> **